### PR TITLE
protocols: Implement Direct Connection Upgrade through Relay (dcutr)

### PR DESCRIPTION
**Pull request by mxinden**
This pull request implements the [Direct Connection Upgrade through Relay](https://github.com/libp2p/specs/pull/173) specification. For the greater picture please see https://github.com/libp2p/rust-libp2p/issues/2052 and https://github.com/libp2p/specs/issues/312.

**Status quo**

It builds on top of https://github.com/libp2p/rust-libp2p/pull/2059 and https://github.com/libp2p/rust-libp2p/pull/2066. Please post review comments related to circuit relay v2 and simultaneous open on https://github.com/libp2p/rust-libp2p/pull/2059 and https://github.com/libp2p/rust-libp2p/pull/2066.

### CODE REVIEW

**Comment by mxinden** [comment]
**With the recent changes I was able to punch the first hole into a Firewall :partying_face:** using _Circuit Relay v2_ https://github.com/libp2p/rust-libp2p/pull/2059, _Direct Connection Upgrade Through Relay_ https://github.com/libp2p/rust-libp2p/pull/2076 and _Multistream Select Simultaneous Open_ https://github.com/libp2p/rust-libp2p/pull/2066.

**Comment by dvc94ch** [comment]
Should I open a PR for including libp2p-quic? I'd really love to see that this approach is "extensible". Looking forward to supporting it in ipfs-embed.

**Comment by mxinden** [comment]
> Should I open a PR for including libp2p-quic?

With `libp2p-quic` you mean https://github.com/ipfs-rust/libp2p-quic combining QUIC and Noise, right?

Obviously I think rust-libp2p should support QUIC + TLS (see https://github.com/libp2p/rust-libp2p/pull/1334 and https://github.com/tomaka/libp2p-rs/tree/quiccc-again). Supporting QUIC + Noise in addition sounds good to me, though this would need to be specified in parallel (https://github.com/libp2p/specs/issues/315).

> I'd really love to see that this approach is "extensible".

Same here. Anything in particular that causes your fear of this not being extensible? 

> Looking forward to supporting it in ipfs-embed.

:tada:

**Comment by dvc94ch** [comment]
> Same here. Anything in particular that causes your fear of this not being extensible?

stuff like this tends to have assumptions backed in. for one the relay is implemented as a transport wrapper instead of a stream muxer wrapper.

> With libp2p-quic you mean https://github.com/ipfs-rust/libp2p-quic combining QUIC and Noise, right?

I guess both could be supported using the same codebase. Since a custom quic version was registered and is used, using feature flags both could be enabled at the same time (although that would probably require a small PR to quinn-proto). The code in the PR you mentioned is out of date, needs quite a bit of work to update to the latest libp2p and quinn. In addition it has many todo's in it and after spending some time on it, I felt it would be best to rewrite it from scratch.

**Comment by dvc94ch** [comment]
ah maybe not. libp2p-quic explicitly only supports ed25519 which I guess would be a deal breaker for any libp2p spec.

**Comment by dvc94ch** [comment]
Has this effort stalled? Would be a shame as this is a really important feature for rust-libp2p

**Comment by mxinden** [comment]
> Has this effort stalled? 

Not at all! Unfortunately hole punching requires many pieces. I am currently working on https://github.com/libp2p/rust-libp2p/issues/1896 (1). Next we will need to rework `NetworkBehaviour::DialPeer` to be able to provide a specific set of addresses to be used (2) and specify whether the established connection would act as a listener or as a dialer (3).

Once all these 3 efforts are done, we can proceed here.

> Would be a shame as this is a really important feature for rust-libp2p

Indeed!

**Comment by elenaf9** [comment]
>  Next we will need to rework NetworkBehaviour::DialPeer to be able to provide a specific set of addresses to be used (2)

@mxinden Could you expand a bit more on what exactly  the desired functionality is, and why `NetworkBehaviour::DialAddress` is not sufficient (I guess because it creates an overhead?)?. I remember reading a discussion about this, but I don't remember the details anymore.
If this is something that can be implemented independently of (1), I'd be happy to tackle it.

**Comment by mxinden** [comment]
> > Next we will need to rework NetworkBehaviour::DialPeer to be able to provide a specific set of addresses to be used (2)
> 
> @mxinden Could you expand a bit more on what exactly the desired functionality is, and why `NetworkBehaviour::DialAddress` is not sufficient (I guess because it creates an overhead?)?. I remember reading a discussion about this, but I don't remember the details anymore.

I documented (2) and (3) in https://github.com/libp2p/rust-libp2p/issues/2249 and https://github.com/libp2p/rust-libp2p/issues/2250. @elenaf9 could you take a look at the two issues and follow up with any questions that might arise on the issues themselves?

> If this is something that can be implemented independently of (1), I'd be happy to tackle it.

In my eyes https://github.com/libp2p/rust-libp2p/issues/1896 (1), https://github.com/libp2p/rust-libp2p/issues/2249 (2) and https://github.com/libp2p/rust-libp2p/issues/2250 (3) are best done in sequence to reduce conflicts and thus hard to collaborate on.

@elenaf9 are you interested in (2) and (3) in particular, or interested in helping to land hole punching in general. In case the latter is true, there are many other outstanding things that can be tackled independently.

The number one isolated thing that comes to my mind is the AutoNAT implementation. There is a specification draft https://github.com/libp2p/specs/pull/362 and a stalled attempt https://github.com/libp2p/rust-libp2p/pull/1672. Note that I sugested the same  to @iduartgomez in https://github.com/freenet/locutus/issues/2#issuecomment-913647111, just to avoid duplicate work.

**Comment by elenaf9** [comment]
> I documented (2) and (3) in #2249 and #2250. @elenaf9 could you take a look at the two issues and follow up with any questions that might arise on the issues themselves?

Thanks, with the issues it's much clearer now.

> The number one isolated thing that comes to my mind is the AutoNAT implementation. There is a specification draft libp2p/specs#362 and a stalled attempt #1672. Note that I sugested the same to @iduartgomez in freenet/locutus#2 (comment), just to avoid duplicate work.
 
@iduartgomez did you already put any effort towards the implementation of AutoNAT?
If this is not the case I would start working on a new attempt, based on the specs and the draft from #1672.

**Comment by iduartgomez** [comment]
> @iduartgomez did you already put any effort towards the implementation of AutoNAT?
> If this is not the case I would start working on a new attempt, based on the specs and the draft from #1672.

No, haven't got to it yet, you can start yourself no problem :) Keep us updated!

File: misc/multistream-select/src/protocol.rs [hunk 1]
```diff
@@ -249,7 +303,7 @@ impl<R> MessageIO<R> {
         }
     }
 
-    /// Drops the [`MessageIO`] resource, yielding the underlying I/O stream.
+    /// Draops the [`MessageIO`] resource, yielding the underlying I/O stream.
```
Comment (r-zig) [inline misc/multistream-select/src/protocol.rs]: spelling -> Drops?

File: protocols/dcutr/examples/client.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        let external_addr = Multiaddr::empty()
+            .with(external_ip.parse::<Ipv4Addr>().unwrap().into())
+            .with(Protocol::Tcp(port.parse().unwrap()));
+        swarm.add_external_address(external_addr, AddressScore::Infinite);
+    } else {
+        panic!();
+    }
+
+    if let (Some(mode), Some(addr)) = (std::env::args().nth(4), std::env::args().nth(5)) {
```
Comment (r-zig) [inline protocols/dcutr/examples/client.rs]: Maybe we can change to structopt with description for each parameter as we did in the relay_v1
and then we can write some example of 3 nodes settings together?

Comment (mxinden) [reply protocols/dcutr/examples/client.rs]: Yes for sure, that would be nice. This is a hack at most at the moment. Want to file a pull request against my branch on my fork?

Comment (r-zig) [reply protocols/dcutr/examples/client.rs]: Yes, I will do.

File: protocols/relay/src/v2.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+//! Implementation of the [libp2p circuit relay v2
+//! specification](https://github.com/libp2p/specs/issues/314).
+
+mod message_proto {
+    include!(concat!(env!("OUT_DIR"), "/message_v2.pb.rs"));
+}
+
+pub mod client;
+mod copy_future;
+mod protocol;
```
Comment (canewsin) [inline protocols/relay/src/v2.rs:30]: Make protocol public mod or Export 
UpgradeError in inbound_hop/inbound_hop and two other modules

Comment (canewsin) [reply protocols/relay/src/v2.rs:30]: @mxinden

**Comment by mxinden** [comment]
With https://github.com/libp2p/rust-libp2p/pull/2059 merged and https://github.com/libp2p/rust-libp2p/pull/2363 likely substituting https://github.com/libp2p/rust-libp2p/pull/2066, I am replacing this pull request with https://github.com/libp2p/rust-libp2p/pull/2438 to ease reviewing.

**Pull request closed without merging**